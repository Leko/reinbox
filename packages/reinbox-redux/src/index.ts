export { actions } from "./actions";
export { createReducer } from "./createReducer";
export { default as Provider } from "./ProviderContainer";
