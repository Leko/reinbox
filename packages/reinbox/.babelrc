{
  "presets": [
    [
      "@babel/preset-typescript",
      {
        "exclude": ["transform-regenerator"]
      }
    ],
    "@babel/preset-typescript",
    "@babel/preset-react"
  ]
}
