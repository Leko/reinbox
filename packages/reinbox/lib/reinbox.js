'use strict';

module.exports = reinbox;

function reinbox() {
    // TODO
}
