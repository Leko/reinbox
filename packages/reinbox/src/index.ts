export { InboxProvider as Provider } from "./InboxProvider";
export { Inbox } from "./Inbox";
export { withPublish, WithPublishProps } from "./withPublish";
