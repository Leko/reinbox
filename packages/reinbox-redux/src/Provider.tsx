import * as React from "react";
import { RawProvider, Context } from "reinbox";

export type Props = {
  namespace?: string;
  children: React.ReactNode;
};
type ConnectedProps = Props & Context;

export const Provider: React.SFC<ConnectedProps> = ({
  children,
  state,
  dismiss,
  publish
}: ConnectedProps) => {
  return (
    <RawProvider
      value={{
        state,
        dismiss,
        publish
      }}
    >
      {children}
    </RawProvider>
  );
};
Provider.defaultProps = {
  namespace: "reinbox"
};
