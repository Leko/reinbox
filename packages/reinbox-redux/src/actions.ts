import { Payload } from "reinbox";

type PublishAction = {
  type: "reinbox/publish";
  payload: {
    topic: string;
    payload: Payload;
  };
};
type DismissAction = {
  type: "reinbox/dismiss";
  payload: {
    topic: string;
    id: string;
  };
};
export type Actions = PublishAction | DismissAction;

const publish = (topic: string, payload: Payload): PublishAction => ({
  type: "reinbox/publish",
  payload: {
    topic,
    payload
  }
});
const dismiss = (topic: string, id: string): DismissAction => ({
  type: "reinbox/dismiss",
  payload: {
    topic,
    id
  }
});

export const actions = {
  publish,
  dismiss
};
