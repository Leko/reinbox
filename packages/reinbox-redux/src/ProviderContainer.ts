import { Dispatch } from "redux";
import { connect } from "react-redux";
import { State, Payload } from "reinbox";
import { actions, Actions } from "./actions";
import { Provider, Props } from "./Provider";

export const mapStateToProps = (
  state: { [key: string]: Object },
  ownProps: Props
) => {
  const { namespace } = ownProps;
  if (!namespace || !state[namespace]) {
    throw new Error(`key ${namespace} not found in Redux state`);
  }

  return {
    state: state[namespace] as State
  };
};

export const mapDispatchToProps = (dispatch: Dispatch<Actions>) => ({
  publish(topic: string, payload: Payload) {
    dispatch(actions.publish(topic, payload));
  },
  dismiss(topic: string, id: string) {
    dispatch(actions.dismiss(topic, id));
  }
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Provider);
