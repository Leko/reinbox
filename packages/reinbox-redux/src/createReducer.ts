// @ts-ignore power-assign not provided DTS
import { assign } from "power-assign";
import { State } from "reinbox";
import { Actions } from "./actions";

export const createReducer = () => {
  const initialState: State = {
    topics: {}
  };

  const reducer = (state: State = initialState, action: Actions) => {
    switch (action.type) {
      case "reinbox/publish": {
        const { topic, payload } = action.payload;
        return assign(state, {
          [`topics.${topic}.${payload.id}`]: payload
        });
      }
      case "reinbox/dismiss": {
        const { topic, id } = action.payload;
        return assign(state, {
          $unset: {
            [`topics.${topic}.${id}`]: ""
          }
        });
      }
      default:
        return state;
    }
  };

  return reducer;
};
