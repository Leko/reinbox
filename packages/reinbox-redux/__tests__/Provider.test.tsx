import * as React from "react";
import * as renderer from "react-test-renderer";
import { Inbox } from "reinbox";
import { Provider } from "../src/Provider";

const Child = () => (
  <Inbox topic="test">
    {({ payload, dismiss }) =>
      payload && (
        <div>
          <span>{payload.id}</span>
          <button onClick={dismiss}>dismiss</button>
        </div>
      )
    }
  </Inbox>
);

const mockProps = {
  dismiss: () => {},
  publish: () => {}
};

test("snapshot with one message", () => {
  const topics = {
    reinbox: {
      xxx: { id: "xxx" }
    }
  };
  const component = renderer.create(
    <Provider {...mockProps} state={{ topics }}>
      <Child />
    </Provider>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test("snapshot with one message and custom namespace", () => {
  const namespace = "ns";
  const topics = {
    [namespace]: {
      xxx: { id: "xxx" }
    }
  };
  const component = renderer.create(
    <Provider {...mockProps} state={{ topics }} namespace={namespace}>
      <Child />
    </Provider>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test("snapshot with no messages", () => {
  const component = renderer.create(
    <Provider {...mockProps} state={{ topics: {} }}>
      <Child />
    </Provider>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
