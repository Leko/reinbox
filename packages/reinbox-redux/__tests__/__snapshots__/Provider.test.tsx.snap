// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot with no messages 1`] = `null`;

exports[`snapshot with one message 1`] = `null`;

exports[`snapshot with one message and custom namespace 1`] = `null`;
